src/backend/distributed/utils/node_metadata.c

```diff
@@ -264,52 +273,160 @@ master_remove_node(PG_FUNCTION_ARGS)
 Datum
 master_disable_node(PG_FUNCTION_ARGS)
 {
-	const bool onlyConsiderActivePlacements = true;
 	text *nodeNameText = PG_GETARG_TEXT_P(0);
 	int32 nodePort = PG_GETARG_INT32(1);
-
 	char *nodeName = text_to_cstring(nodeNameText);
+	WorkerNode *workerNode = ModifiableWorkerNode(nodeName, nodePort);
 	bool isActive = false;
+	bool onlyConsiderActivePlacements = false;
 
-	WorkerNode *workerNode = NULL;
+	if (WorkerNodeIsPrimary(workerNode))
+	{
+		/* Delete refrence table placements so they are not taken into account for
+		 * the check if there are placements after this */
+		DeleteAllReferenceTablePlacementsFromNodeGroup(workerNode->groupId);
 
-	CheckCitusVersion(ERROR);
+		if (NodeGroupHasShardPlacements(workerNode->groupId,
+										onlyConsiderActivePlacements))
+		{
+			ereport(NOTICE, (errmsg(
+								 "Node %s:%d has active shard placements. Some queries "
+								 "may fail after this operation. Use "
+								 "SELECT master_activate_node('%s', %d) to activate this "
+								 "node back.",
+								 workerNode->workerName, nodePort, workerNode->workerName,
+								 nodePort)));
+		}
+	}
 
-	EnsureCoordinator();
+	SetNodeState(nodeName, nodePort, isActive);
 
-	/* take an exclusive lock on pg_dist_node to serialize pg_dist_node changes */
-	LockRelationOid(DistNodeRelationId(), ExclusiveLock);
+	CleanUpReferenceTables(workerNode);
 
-	workerNode = FindWorkerNodeAnyCluster(nodeName, nodePort);
-	if (workerNode == NULL)
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * master_set_node_property sets a property of the node
+ */
+Datum
+master_set_node_property(PG_FUNCTION_ARGS)
+{
+	text *nodeNameText = PG_GETARG_TEXT_P(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+	text *propertyText = PG_GETARG_TEXT_P(2);
+	bool value = PG_GETARG_BOOL(3);
+
+	WorkerNode *workerNode = ModifiableWorkerNode(text_to_cstring(nodeNameText),
+												  nodePort);
+
+	if (strcmp(text_to_cstring(propertyText), "shouldhaveshards") == 0)
 	{
-		ereport(ERROR, (errmsg("node at \"%s:%u\" does not exist", nodeName, nodePort)));
+		SetShouldHaveShards(workerNode, value);
 	}
+	else
+	{
+		ereport(ERROR, (errmsg(
+							"It's only supported to set 'shouldhaveshards' using this function"
```

Comment:
This sentence is a little strange.
"only the 'shouldhaveshards' property can be set using this function"